src/request-pipeline/file-request.ts

```diff
@@ -5,34 +5,71 @@ import { EventEmitter } from 'events';
 import { parse } from 'url';
 import { MESSAGE, getText } from '../messages';
 import { stat, access } from '../utils/promisified-functions';
+import asar from 'asar';
+import { toReadableStream } from '../utils/buffer';
 
-const DISK_RE: RegExp = /^\/[A-Za-z]:/;
+const DISK_RE: RegExp           = /^\/[A-Za-z]:/;
+const ASAR_ARCHIVE_PATH: RegExp = /^.*\.asar/;
 
-const TARGET_IS_NOT_FILE = 'The target of the operation is not a file';
+const TARGET_IS_NOT_FILE              = 'The target of the operation is not a file';
+const ASAR_ARCHIVE_TARGET_IS_NOT_FILE = 'The asar archive target of the operation is not a file';
 
 export default class FileRequest extends EventEmitter {
     url: string;
     path: string;
+    asarArchivePath: string;
 
     constructor (url: string) {
         super();
 
-        this.url  = url;
-        this.path = FileRequest._getPath(url);
+        this.url             = url;
+        this.path            = FileRequest._getPath(url);
+        this.asarArchivePath = FileRequest._getAsarArchivePath(this.path);
 
         this._initEvents();
     }
 
     _initEvents () {
-        stat(this.path)
-            .then((stats: fs.Stats) => {
-                if (!stats.isFile())
-                    throw new Error(TARGET_IS_NOT_FILE);
-
-                return access(this.path, fs.constants.R_OK);
-            })
-            .then(() => this._onOpen())
-            .catch((err: Error) => this._onError(err));
+        if (this.asarArchivePath) {
+            stat(this.asarArchivePath)
+                .catch(() => {
+                    throw new Error(ASAR_ARCHIVE_TARGET_IS_NOT_FILE);
+                })
+                .then(() => access(this.asarArchivePath, fs.constants.R_OK))
+                .then(() => this._onOpen())
+                .catch((err: Error) => {
+
+                    return this._onError(err);
+                });
+        }
+        else {
+            stat(this.path)
+                .then((stats: fs.Stats) => {
+                    if (!stats.isFile())
+                        throw new Error(TARGET_IS_NOT_FILE);
+
+                    return access(this.path, fs.constants.R_OK);
+                })
+                .then(() => this._onOpen())
+                .catch((err: Error) => this._onError(err));
+        }
+    }
+
+    static _getAsarArchivePath (path: string) : string {
```

Comment:
Move all ASAR related code to the separate file and the following method signatures:
```js
asar.getArchivePath(...)
asar.getFilePathInArchive(...)
asar.getReadStream(...)
```